### ISSUE TITLE
windows_2016_docker.json produces 0 byte vmdk

### ISSUE THREAD

**Issue by asbjornu** [StefanScherer/packer-windows#87]
I'm trying to get [Octopus Deploy](https://hub.docker.com/r/octopusdeploy/octopusdeploy/) run in Docker (version `18.03.0-ce-rc1, build c160c73`) on my macOS 10.12.6. I've installed VirtualBox (version `5.2.8r121009`), downloaded the Windows 2016 ISO and seemingly successfully executed the following command:

```shell
packer build --only=virtualbox-iso -var 'iso_url=/Users/bitbear/Downloads/14393.0.161119-1705.RS1_REFRESH_SERVER_EVAL_X64FRE_EN-US.ISO' windows_2016_docker.json
```

After a few hours, this command finishes with the following:

```shell
Build 'virtualbox-iso' finished.

==> Builds finished. The artifacts of successful builds are:
--> virtualbox-iso: 'virtualbox' provider box: windows_2016_docker_virtualbox.box
```

However, when I execute the following command, a `.box` file is created with a `WindowsServer2016Docker-disk001.vmdk` file in it that is 0 bytes in size:

```shell
vagrant box add windows_2016_docker windows_2016_docker_virtualbox.box
```

After copying `vagrantfile-windows_2016.template` to `Vagrantfile` and modifying it according to work with my setup, `vagrant up` fails with the following:

```
There was an error while executing `VBoxManage`, a CLI used by Vagrant
for controlling VirtualBox. The command and stderr is shown below.

Command: ["import", "/Users/bitbear/.vagrant.d/boxes/windows_2016_docker/0/virtualbox/box.ovf", "--vsys", "0", "--vmname", "WindowsServer2016Docker_1519934437898_19158", "--vsys", "0", "--unit", "8", "--disk", "/Users/bitbear/VirtualBox VMs/WindowsServer2016Docker_1519934437898_19158/WindowsServer2016Docker-disk001.vmdk"]

Stderr: 0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
Interpreting /Users/bitbear/.vagrant.d/boxes/windows_2016_docker/0/virtualbox/box.ovf...
OK.
0%...
Progress state: VBOX_E_FILE_ERROR
VBoxManage: error: Appliance import failed
VBoxManage: error: Could not open the medium storage unit '/Users/bitbear/.vagrant.d/boxes/windows_2016_docker/0/virtualbox/WindowsServer2016Docker-disk001.vmdk'.
VBoxManage: error: VMDK: error reading the magic number in '/Users/bitbear/.vagrant.d/boxes/windows_2016_docker/0/virtualbox/WindowsServer2016Docker-disk001.vmdk' (VERR_EOF).
VBoxManage: error: VD: error VERR_VD_VMDK_INVALID_HEADER opening image file '/Users/bitbear/.vagrant.d/boxes/windows_2016_docker/0/virtualbox/WindowsServer2016Docker-disk001.vmdk' (VERR_VD_VMDK_INVALID_HEADER)
VBoxManage: error: Details: code VBOX_E_FILE_ERROR (0x80bb0004), component ApplianceWrap, interface IAppliance
VBoxManage: error: Context: "RTEXITCODE handleImportAppliance(HandlerArg *)" at line 886 of file VBoxManageAppliance.cpp
```

When I inspect the file `/Users/bitbear/.vagrant.d/boxes/windows_2016_docker/0/virtualbox/WindowsServer2016Docker-disk001.vmdk`, it is indeed an empty file of 0 bytes in size. Any ideas what might be causing this and how I can fix it?

**Comment by StefanScherer** [comment]
Which version of Packer do you use? 
My latest builds are done with Packer 1.1.3. 
Which size is the box file? It should be about 10 GByte.

**Comment by basictheprogram** [comment]
Do you have enough free disk space to hold the 10 GByte file?

**Comment by asbjornu** [comment]
I'm using version `1.2.1` of packer and I have 160 GB of free disk space. The box is 10.89 GB.

**Comment by StefanScherer** [comment]
The box size looks good. Can you list the content of the box file?
It should look like this:

```
$ .\tar.exe tvf .\windows_2016_docker_virtualbox.box
-rw-rw-rw-  0 0      0        2347 Jan 23 01:43 Vagrantfile
-rw-rw-rw-  0 0      0 10767669248 Jan 23 01:43 WindowsServer2016Docker-disk001.vmdk
-rw-rw-rw-  0 0      0       10416 Jan 23 01:43 box.ovf
-rw-rw-rw-  0 0      0          26 Jan 23 01:43 metadata.json
```

**Comment by asbjornu** [comment]
```shell
$ tar tvf windows_2016_docker_virtualbox.box
-rw-r--r--  0 bitbear staff    2286 Mar  1 20:50 Vagrantfile
-rw-r--r--  0 bitbear staff       0 Mar  1 20:50 WindowsServer2016Docker-disk001.vmdk
-rw-r--r--  0 bitbear staff   10412 Mar  1 20:50 box.ovf
-rw-r--r--  0 bitbear staff      26 Mar  1 20:50 metadata.json
```

**Comment by StefanScherer** [comment]
Well, I don't know where the error during the packer build happened to produce this corrupt box file.
I would suggest to just rebuild the box and see if it works. You could also use the [`StefanScherer/windows_2016_docker`](https://app.vagrantup.com/StefanScherer/boxes/windows_2016_docker) box from Vagrant cloud.

**Comment by asbjornu** [comment]
I’ve rebuilt the box four times due to other hints I’ve found around the web of what the problem might be. Nothing has helped, which is why I ended up asking here.

I’ll definitely try your box from Vagrant cloud, thanks!

**Comment by basictheprogram** [comment]
Can duplicate.

macOS 10.13.3
$ packer --version
1.2.1

$ vboxmanage --version
5.2.8r121009

```shell
$ ls -lh *.box
-rw-r--r--  1 tanner  staff    10G Mar  4 03:30 windows_2016_docker_virtualbox.box
```
The .box is 10G in size

```shell
$ tar -tvf windows_2016_docker_virtualbox.box
-rw-r--r--  0 tanner staff    2286 Mar  4 03:25 Vagrantfile
-rw-r--r--  0 tanner staff       0 Mar  4 03:25 WindowsServer2016Docker-disk001.vmdk

-rw-r--r--  0 tanner staff   10412 Mar  4 03:25 box.ovf
-rw-r--r--  0 tanner staff      26 Mar  4 03:25 metadata.json
```
But the .vmdk is zero?

Where is the 10G of disk space coming from?

**Comment by asbjornu** [comment]
Yes, I'm using HFS+ since I have a Fusion drive. I'm waiting for [Fusion support in APFS](https://www.macrumors.com/2017/09/25/apfs-fusion-drive-high-sierra-update/) before I convert.

**Comment by basictheprogram** [comment]
HFS+ has  Max. file size | 8 exabyte

**Comment by StefanScherer** [comment]
Interesting, so HFS+ does not have a problem with the box file size and extracting the contents.
So it seems a problem while creating the tar file.
You can run Packer with debug output
```
export PACKER_LOG=1
packer build ...
```
To see which command is called during the box file creation.
I don't know if it's a built-in command in the packer binary or if some external tools from macOS are called.

**Comment by basictheprogram** [comment]
Not seeing anything obvious in the logs.


[typescript.txt](https://github.com/StefanScherer/packer-windows/files/1782300/typescript.txt)

**Comment by basictheprogram** [comment]
"keep_input_artifact": true
HFS+ filesystem
[Packer Logs](https://github.com/StefanScherer/packer-windows/files/1783774/typescript.txt)

output-virtualbox-iso/
```shell
$ ls -l
total 21220296
-rw-r--r--  1 tanner  staff  10864777216 Mar  5 18:46 WindowsServer2016Docker-disk001.vmdk
-rwx------  1 tanner  staff        10412 Mar  5 18:23 WindowsServer2016Docker.ovf
```
```shell
$ ls -l *.box
-rw-r--r--  1 tanner  staff  10729788371 Mar  5 18:56 windows_2016_docker_virtualbox.box
```
```shell
$ tar -tvf windows_2016_docker_virtualbox.box
-rw-r--r--  0 tanner staff    2286 Mar  5 18:51 Vagrantfile
-rw-r--r--  0 tanner staff       0 Mar  5 18:51 WindowsServer2016Docker-disk001.vmdk
-rw-r--r--  0 tanner staff   10412 Mar  5 18:51 box.ovf
-rw-r--r--  0 tanner staff      26 Mar  5 18:51 metadata.json
```

**Comment by StefanScherer** [comment]
Thanks. This really looks like a problem inside packer

```
2018/03/05 18:51:25 packer: 2018/03/05 18:51:25 Turning dir into box: /var/folders/5m/b4xyt3kj2sx8sbrksf0c6q0w0000gn/T/packer821802642 => windows_2016_docker_virtualbox.box
2018/03/05 18:51:25 packer: 2018/03/05 18:51:25 Compressing with gzip compression level: -1
2018/03/05 18:51:26 packer: 2018/03/05 18:51:26 Skipping directory '/var/folders/5m/b4xyt3kj2sx8sbrksf0c6q0w0000gn/T/packer821802642' for box 'windows_2016_docker_virtualbox.box'
2018/03/05 18:51:26 packer: 2018/03/05 18:51:26 Box add: '/var/folders/5m/b4xyt3kj2sx8sbrksf0c6q0w0000gn/T/packer821802642/Vagrantfile' to 'windows_2016_docker_virtualbox.box'
2018/03/05 18:51:26 ui: [0;32m    virtualbox-iso (vagrant): Compressing: Vagrantfile[0m
[0;32m    virtualbox-iso (vagrant): Compressing: Vagrantfile[0m
2018/03/05 18:51:26 packer: 2018/03/05 18:51:26 Box add: '/var/folders/5m/b4xyt3kj2sx8sbrksf0c6q0w0000gn/T/packer821802642/WindowsServer2016Docker-disk001.vmdk' to 'windows_2016_docker_virtualbox.box'
2018/03/05 18:51:26 ui: [0;32m    virtualbox-iso (vagrant): Compressing: WindowsServer2016Docker-disk001.vmdk[0m
[0;32m    virtualbox-iso (vagrant): Compressing: WindowsServer2016Docker-disk001.vmdk[0m
[0;32m    virtualbox-iso (vagrant): Compressing: box.ovf[0m
[0;32m    virtualbox-iso (vagrant): Compressing: metadata.json[0m
2018/03/05 18:56:09 packer: 2018/03/05 18:56:09 Box add: '/var/folders/5m/b4xyt3kj2sx8sbrksf0c6q0w0000gn/T/packer821802642/box.ovf' to 'windows_2016_docker_virtualbox.box'
2018/03/05 18:56:09 ui: [0;32m    virtualbox-iso (vagrant): Compressing: box.ovf[0m
2018/03/05 18:56:09 packer: 2018/03/05 18:56:09 Box add: '/var/folders/5m/b4xyt3kj2sx8sbrksf0c6q0w0000gn/T/packer821802642/metadata.json' to 'windows_2016_docker_virtualbox.box'
2018/03/05 18:56:09 ui: [0;32m    virtualbox-iso (vagrant): Compressing: metadata.json[0m
2018/03/05 18:56:09 [INFO] (telemetry) ending vagrant
```

The tar file is created internally: https://github.com/hashicorp/packer/blob/4d3a762e85e7e8050134a4e33e1f87a54029dcb8/post-processor/vagrant/util.go#L71-L155
but this function has no recent changes.

**Comment by basictheprogram** [comment]
Can duplicate this problem with macOS sierra (10.12.6) and HFS+ filesystem.  

Time to open a bug in the packer project?

**Comment by StefanScherer** [comment]
There's one config option `compression_level` which can be set to `0` to turn off compression. Don't know if that makes a difference.

Issue with Golang or one of the imported packages?, maybe.
Maybe create a small golang program from the DirToBox function, removing the ui parameter and run it with some big files.

An issue in packer project may help to get some feedback. But I haven't seen a similar issue there yet, so probably most people work with much smaller vmdk sizes.

**Comment by StefanScherer** [comment]
@asbjornu With `docker-compose` you only can start containers on one Docker host. If you want to mix Linux and Windows workloads you need eg. Docker Swarm running a Linux and Windows node and then you can use a docker-compose.yml to deploy a stack.

With later Windows Server like 1709, 1803, ... and latest nightly Docker builds you may use LCOW to start Linux containers directly on the Windows Docker host. But this works only with HyperV enabled in the VM, but VirtualBox does not have nested virtualization feature to do that.

**Comment by asbjornu** [comment]
@StefanScherer: Thanks. As I write in https://github.com/docker/compose/issues/3987#issuecomment-370616369 I finally figured this out as well. I hoped to not have to start from scratch with Docker Swarm, since I've never used it before and really enjoy the simplicity of Docker Compose. I might as well just go with Kubernetes then, since I know that better than Docker Swarm, but it too adds infinite amounts of complexity I don't really want to deal with in this very simple demo project.

> With later Windows Server like 1709, 1803, ... and latest nightly Docker builds you may use LCOW to start Linux containers directly on the Windows Docker host. But this works only with HyperV enabled in the VM, but VirtualBox does not have nested virtualization feature to do that.

Yeah, I considered that but VirtualBox makes that impossible. So I don't really know what to do. I think I might just manually run Octopus Deploy within a Windows VM and skip Docker for this purpose altogether.

**Comment by StefanScherer** [comment]
The workaround for now: Use Packer 1.1.3

I just rebuilt a VirtualBox box file (`brew switch packer 1.1.3`) and everything looks fine:

```
$ tar tvf windows_2016_docker_virtualbox.box 
-rw-r--r--  0 501    20        2286 Mar 11 14:27 Vagrantfile
-rw-r--r--  0 501    20 10952051712 Mar 11 14:27 WindowsServer2016Docker-disk001.vmdk
-rw-r--r--  0 501    20       10412 Mar 11 14:27 box.ovf
-rw-r--r--  0 501    20          26 Mar 11 14:27 metadata.json
~/code/packer-windows on my*
$ ls -l *box
-rw-r--r--  1 stefan  staff  10817278351 Mar 11 14:28 windows_2016_docker_virtualbox.box
```

**Comment by StefanScherer** [comment]
Created https://github.com/golang/go/issues/24599 for further investigation.

**Comment by StefanScherer** [comment]
@basictheprogram Your fix/workaround includes using the homebrew GNU tar and replacing the embedded tar in Vagrant?

**Comment by StefanScherer** [comment]
Weird. But as long as it works for you. 
It will be officially fixed with Packer 1.2.3, PR https://github.com/hashicorp/packer/pull/6084 got merged.
So using Packer 1.1.3 or 1.2.3 fixes the problem.
